pkgs/games/factorio/default.nix

```diff
@@ -4,80 +4,108 @@
 # Begin download parameters
 , username ? ""
 , password ? ""
+, releaseType
 }:
 
+assert releaseType == "alpha" || releaseType == "headless";
+
+with stdenv.lib;
 let
   version = "0.12.28";
+  isHeadless = releaseType == "headless";
+
+  arch = if stdenv.system == "x86_64-linux" then {
+    inUrl = "linux64";
+    inTar = "x64";
+  } else if stdenv.system == "i686-linux" then {
+    inUrl = "linux32";
+    inTar = "i386";
+  } else abort "Unsupported platform";
 
-  fetch = callPackage ./fetch.nix { username = username; password = password; };
-  arch = if stdenv.system == "x86_64-linux" then "x64"
-         else if stdenv.system == "i686-linux" then "x32"
-         else abort "Unsupported platform";
+  authenticatedFetch = callPackage ./fetch.nix { inherit username password; };
 
-  variants = {
+  fetch = rec {
+    url = "https://www.factorio.com/get-download/${version}/${releaseType}/${arch.inUrl}";
+    name = "factorio_${releaseType}_${arch.inTar}-${version}.tar.gz"; # TODO take this from 302 redirection somehow? fetchurl doesn't help.
     x64 = {
-      url = "https://www.factorio.com/get-download/${version}/alpha/linux64";
-      sha256 = "01si5n9hb2h0c5q8k3hr3nphsakp9kki84qyp70dgddwqsn8wfjl";
+      alpha = authenticatedFetch { inherit      url; sha256 = "0vngfrjjib99k6czhg32rikfi36i3p3adx4mxc1z8bi5n70dbwqb"; };
```

Comment:
Try to keep lines below 80 characters when feasible. You can split these across multiple lines.